src/Modules/UngradedModules.jl

```diff
@@ -0,0 +1,3653 @@
+  N = parent(v)
+  if N===M
+    return v
+  end
+
+  # Breadth-First Search to find path to N:
+  parent_hom = IdDict{SubQuo,ModuleMap}()
+  modules = [M]
+  found_N = false
+  for A in modules
+    for H in A.incoming_morphisms
+      B = domain(H)
+      if B!==A # on trees "B!==A" is enough!
+        if findfirst(x->x===B,modules) == nothing #if !(B in modules) doesn't work since it uses == instead of ===
+          parent_hom[B] = H
+          push!(modules,B)
+        end
+      end
+      if B===N
+        found_N = true
+        break
+      end
+    end
+    if found_N
+      break
+    end
+  end
+  if !found_N
+    throw(DomainError("There is no path of canonical homomorphisms between the modules!"))
+  end
+  result = v
+  A = N
+  while A !== M
+    H = parent_hom[A]
+    result = H(result)
+    A = codomain(H)
+  end
+  return result
+end
+
+function all_canonical_maps(M::SubQuo, N::SubQuo)
+  # from N to M
+
+  all_paths = []
+
+  function helper_dfs!(U::SubQuo, D::SubQuo, visited::Vector{<:ModuleMap}, path::Vector)
+    if U === D
+      push!(all_paths, path)
+      return
+    end
+    for neighbor_morphism in U.outgoing_morphisms
+      if findfirst(x->x===neighbor_morphism, visited) === nothing #if !(neighbor_morphism in visited) doesn't work since it uses == instead of ===
+        helper_dfs!(codomain(neighbor_morphism), D, vcat(visited, [neighbor_morphism]), union(path, [neighbor_morphism]))
+      end
+    end
+  end
+
+  helper_dfs!(N, M, Vector{ModuleMap}(), [])
+
+  morphisms = Vector{ModuleMap}()
+  for path in all_paths
+    phi = identity_map(N)
+    for h in path
+      phi = phi*h
+    end
+    push!(morphisms, phi)
+  end
+
+  return morphisms
+end
+
+#############################
+# Useful functions
+#############################
+
+@doc Markdown.doc"""
+    register_morphism!(f::ModuleMap)
+
+Cache the morphism `f` in the corresponding caches of the domain and codomain of `f`.
+"""
+function register_morphism!(f::ModuleMap)
+  push!(domain(f).outgoing_morphisms, f)
+  push!(codomain(f).incoming_morphisms, f)
+end
+
+#############################
+#TODO move to Hecke
+#  re-evaluate and use or not
+function differential(c::Hecke.ChainComplex, i::Int)
+  return map(c,length(c)-i)
+end
+
+function module_in_complex(c::Hecke.ChainComplex, i::Int)
+  return obj(c,length(c)-i)
+end
+
+getindex(c::Hecke.ChainComplex, i::Int) = module_in_complex(c,i)
+
+function Base.getindex(r::Hecke.SRow, u::UnitRange)
+  R = base_ring(r)
+  s = sparse_row(R)
+  shift = 1-first(u)
+  for (p,v) = r
+    if p in u
+      push!(s.pos, p+shift)
+      push!(s.values, v)
+    end
+  end
+  return s
+end
+
+function Base.getindex(r::Hecke.SRow, R::AbstractAlgebra.Ring, u::UnitRange)
+  s = sparse_row(R)
+  shift = 1-first(u)
+  for (p,v) = r
+    if p in u
+      push!(s.pos, p+shift)
+      push!(s.values, v)
+    end
+  end
+  return s
+end
+
+function getindex(a::Hecke.SRow, b::AbstractArray{Int, 1})
```

Comment:
```suggestion
function getindex(a::Hecke.SRow, b::AbstractVector{Int})
```
Please, where possible replace uses of `Array` by `Vector` or `Matrix`